internal/pkg/cli/pipeline_list.go

```diff
@@ -0,0 +1,146 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package cli
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/aws/copilot-cli/internal/pkg/aws/codepipeline"
+	"github.com/aws/copilot-cli/internal/pkg/aws/sessions"
+	"github.com/aws/copilot-cli/internal/pkg/config"
+	"github.com/aws/copilot-cli/internal/pkg/deploy"
+	"github.com/aws/copilot-cli/internal/pkg/term/prompt"
+	"github.com/aws/copilot-cli/internal/pkg/term/selector"
+	"github.com/spf13/cobra"
+)
+
+const (
+	pipelineListAppNamePrompt = "Which application are the pipelines in?"
+	pipelineListAppNameHelper = "An application is a collection of related services."
+)
+
+type listPipelineVars struct {
+	appName          string
+	shouldOutputJSON bool
+}
+
+type listPipelineOpts struct {
+	listPipelineVars
+	pipelineSvc pipelineGetter
+	prompt      prompter
+	sel         configSelector
+	w           io.Writer
+}
+
+func newListPipelinesOpts(vars listPipelineVars) (*listPipelineOpts, error) {
+	store, err := config.NewStore()
+	if err != nil {
+		return nil, err
+	}
+
+	defaultSession, err := sessions.NewProvider().Default()
+	if err != nil {
+		return nil, fmt.Errorf("default session: %w", err)
+	}
+	prompter := prompt.New()
+	return &listPipelineOpts{
+		listPipelineVars: vars,
+		pipelineSvc:      codepipeline.New(defaultSession),
+		prompt:           prompter,
+		sel:              selector.NewConfigSelect(prompter, store),
+		w:                os.Stdout,
+	}, nil
+}
+
+// Ask asks for fields that are required but not passed in.
+func (o *listPipelineOpts) Ask() error {
+	if o.appName != "" {
+		return nil
+	}
+	app, err := o.sel.Application(pipelineListAppNamePrompt, pipelineListAppNameHelper)
+	if err != nil {
+		return fmt.Errorf("select application: %w", err)
+	}
+	o.appName = app
+	return nil
+}
+
+// Execute writes the pipelines.
+func (o *listPipelineOpts) Execute() error {
+	var out string
+	if o.shouldOutputJSON {
```

Comment:
Does this fit into the paradigm we started with the `job ls` command?  Could we move this logic to the `list` package? https://github.com/aws/copilot-cli/blob/mainline/internal/pkg/list/list.go It looks like this uses a lot of the same APIs and flags as the other list commands so it might go nicely.

Reply:
I was modeling this off of the `env list` command, so wasn't aware of ongoing restructuring.